// @eltkof7/webimg-cli/dist/lib.js
// src/convert.ts
import sharp from "sharp";
import fs2 from "fs";
import path3 from "path";
import os from "os";

// src/cache.ts
import fs from "fs";
import path from "path";
import crypto from "crypto";
var CACHE_FILE = ".webimg-cache.json";
var ConversionCache = class _ConversionCache {
  data = {};
  file;
  enabled;
  cwd;
  constructor(cwd, enabled = true) {
    this.cwd = cwd;
    this.file = path.join(cwd, CACHE_FILE);
    this.enabled = enabled;
    if (enabled && fs.existsSync(this.file)) {
      try {
        this.data = JSON.parse(fs.readFileSync(this.file, "utf8"));
      } catch {
        this.data = {};
      }
    }
  }
  static hashFile(filePath) {
    const buf = fs.readFileSync(filePath);
    return crypto.createHash("sha1").update(buf).digest("hex");
  }
  static optionsHash(cfg) {
    const payload = {
      version: 1,
      formats: cfg.formats,
      quality: cfg.quality,
      resize: cfg.resize,
      maxWidth: cfg.maxWidth,
      responsive: cfg.responsive,
      suffix: cfg.suffix,
      keepMetadata: cfg.keepMetadata
    };
    return crypto.createHash("sha1").update(JSON.stringify(payload)).digest("hex");
  }
  key(filePath) {
    const abs = path.resolve(this.cwd, filePath);
    const rel = path.relative(this.cwd, abs);
    if (rel && !rel.startsWith("..") && !path.isAbsolute(rel)) return path.normalize(rel);
    return path.normalize(abs);
  }
  absolute(filePath) {
    return path.resolve(this.cwd, filePath);
  }
  hasEntry(file, outputs) {
    if (!this.enabled) return false;
    const entry = this.data[this.key(file)];
    if (!entry) return false;
    const normalizedOutputs = outputs.map((o) => this.key(o));
    return normalizedOutputs.every((o) => entry.outputs.includes(o));
  }
  isUpToDate(file, outputs, optionsHash) {
    if (!this.enabled) return false;
    const entry = this.data[this.key(file)];
    if (!entry) return false;
    if (entry.optionsHash !== optionsHash) return false;
    const normalizedOutputs = outputs.map((o) => this.key(o));
    if (!normalizedOutputs.every((o) => entry.outputs.includes(o) && fs.existsSync(this.absolute(o)))) return false;
    const hash = _ConversionCache.hashFile(this.absolute(file));
    return hash === entry.hash;
  }
  record(file, outputs, optionsHash) {
    if (!this.enabled) return;
    this.data[this.key(file)] = {
      hash: _ConversionCache.hashFile(this.absolute(file)),
      outputs: outputs.map((o) => this.key(o)),
      optionsHash,
      mtime: Date.now()
    };
  }
  save() {
    if (!this.enabled) return;
    fs.writeFileSync(this.file, JSON.stringify(this.data, null, 2));
  }
};

// src/paths.ts
import path2 from "path";
function resolveInputRoot(cwd, cfg) {
  return path2.resolve(cwd, cfg.input || ".");
}
function resolveOutputRoot(cwd, cfg) {
  return cfg.output ? path2.resolve(cwd, cfg.output) : null;
}
function isSubpath(parent, child) {
  const rel = path2.relative(parent, child);
  return rel === "" || !rel.startsWith("..") && !path2.isAbsolute(rel);
}
function toPosixPath(p) {
  return p.split(path2.sep).join("/");
}
function toDisplayPath(cwd, file) {
  const abs = path2.resolve(cwd, file);
  const rel = path2.relative(cwd, abs);
  if (rel && !rel.startsWith("..") && !path2.isAbsolute(rel)) return path2.normalize(rel);
  if (rel === "") return ".";
  return path2.normalize(abs);
}
function relativeDirFromInput(sourcePath, cwd, cfg) {
  const sourceDir = path2.dirname(sourcePath);
  const inputRoot = resolveInputRoot(cwd, cfg);
  const relFromInput = path2.relative(inputRoot, sourceDir);
  if (relFromInput === "") return "";
  if (!relFromInput.startsWith("..") && !path2.isAbsolute(relFromInput)) return relFromInput;
  const relFromCwd = path2.relative(cwd, sourceDir);
  if (relFromCwd === "" || !relFromCwd.startsWith("..") && !path2.isAbsolute(relFromCwd)) return relFromCwd;
  return "";
}
function planOutputs(source, cfg, cwd) {
  const sourcePath = path2.resolve(cwd, source);
  const ext = path2.extname(sourcePath);
  const baseName = path2.basename(sourcePath, ext);
  const outputRoot = resolveOutputRoot(cwd, cfg);
  const dir = outputRoot ? path2.join(outputRoot, relativeDirFromInput(sourcePath, cwd, cfg)) : path2.dirname(sourcePath);
  const densities = cfg.responsive.length ? cfg.responsive : [1];
  const results = [];
  for (const fmt of cfg.formats) {
    for (const d of densities) {
      const dSuffix = densities.length > 1 ? `@${d}x` : "";
      const name = `${baseName}${cfg.suffix}${dSuffix}.${fmt}`;
      results.push({
        output: path2.join(dir, name),
        format: fmt,
        density: d,
        width: cfg.resize ? cfg.resize * d : void 0
      });
    }
  }
  return results;
}
function filterPlannedOutputsFromInputs(files, cfg, cwd) {
  const inputPaths = files.map((file) => path2.resolve(cwd, file));
  const inputSet = new Set(inputPaths);
  const generatedOutputs = /* @__PURE__ */ new Set();
  const outputFormats = new Set(cfg.formats);
  const groups = /* @__PURE__ */ new Map();
  for (const source of inputPaths) {
    const ext = path2.extname(source).slice(1).toLowerCase();
    const base = path2.basename(source, path2.extname(source));
    const key = `${path2.dirname(source)}\0${base}`;
    groups.set(key, [...groups.get(key) ?? [], source]);
    if (outputFormats.has(ext) && /@\d+x$/i.test(base)) generatedOutputs.add(source);
    for (const out of planOutputs(source, cfg, cwd)) {
      const outputPath = path2.resolve(cwd, out.output);
      if (outputPath !== source && inputSet.has(outputPath)) generatedOutputs.add(outputPath);
    }
  }
  for (const group of groups.values()) {
    const hasSourceSibling = group.some((file) => !outputFormats.has(path2.extname(file).slice(1).toLowerCase()));
    if (!hasSourceSibling) continue;
    for (const file of group) {
      const ext = path2.extname(file).slice(1).toLowerCase();
      if (outputFormats.has(ext)) generatedOutputs.add(file);
    }
  }
  return inputPaths.filter((file) => !generatedOutputs.has(file)).map((file) => toDisplayPath(cwd, file));
}

// src/convert.ts
async function applyFormat(pipeline, format, quality) {
  switch (format) {
    case "webp":
      return pipeline.webp({ quality });
    case "avif":
      return pipeline.avif({ quality });
    case "jpeg":
      return pipeline.jpeg({ quality, mozjpeg: true });
    case "png":
      return pipeline.png({ quality, compressionLevel: 9 });
  }
}
async function convertFile(source, cfg, cache, cwd) {
  const sourcePath = path3.resolve(cwd, source);
  const displaySource = toDisplayPath(cwd, sourcePath);
  const outputs = planOutputs(sourcePath, cfg, cwd);
  const outPaths = outputs.map((o) => o.output);
  const optionsHash = ConversionCache.optionsHash(cfg);
  if (cache.isUpToDate(sourcePath, outPaths, optionsHash)) {
    const beforeSize2 = fs2.statSync(sourcePath).size;
    return outputs.map((o) => ({
      source: displaySource,
      output: toDisplayPath(cwd, o.output),
      format: o.format,
      density: o.density,
      status: "cached",
      beforeSize: beforeSize2,
      afterSize: fs2.existsSync(o.output) ? fs2.statSync(o.output).size : 0
    }));
  }
  const beforeSize = fs2.statSync(sourcePath).size;
  const overwriteKnownOutputs = cache.hasEntry(sourcePath, outPaths);
  const results = [];
  for (const out of outputs) {
    const sameAsSource = path3.resolve(out.output) === sourcePath;
    if (sameAsSource) {
      results.push({
        source: displaySource,
        output: toDisplayPath(cwd, out.output),
        format: out.format,
        density: out.density,
        status: "error",
        beforeSize,
        afterSize: 0,
        error: "Le chemin de sortie est identique au fichier source. Utilisez --suffix ou --output."
      });
      continue;
    }
    if (fs2.existsSync(out.output) && !overwriteKnownOutputs && !cfg.dryRun) {
      results.push({
        source: displaySource,
        output: toDisplayPath(cwd, out.output),
        format: out.format,
        density: out.density,
        status: "skipped",
        beforeSize,
        afterSize: fs2.statSync(out.output).size
      });
      continue;
    }
    if (cfg.dryRun) {
      results.push({
        source: displaySource,
        output: toDisplayPath(cwd, out.output),
        format: out.format,
        density: out.density,
        status: "dryrun",
        beforeSize,
        afterSize: 0
      });
      continue;
    }
    try {
      fs2.mkdirSync(path3.dirname(out.output), { recursive: true });
      let p = sharp(sourcePath);
      const meta = await p.metadata();
      let targetWidth = out.width;
      if (cfg.maxWidth && (!targetWidth || targetWidth > cfg.maxWidth)) {
        targetWidth = cfg.maxWidth;
      }
      if (targetWidth && meta.width && targetWidth < meta.width) {
        p = p.resize({ width: targetWidth, withoutEnlargement: true });
      }
      if (cfg.keepMetadata) {
        p = p.withMetadata();
      }
      p = await applyFormat(p, out.format, cfg.quality);
      const tmp = out.output + ".tmp";
      await p.toFile(tmp);
      fs2.renameSync(tmp, out.output);
      results.push({
        source: displaySource,
        output: toDisplayPath(cwd, out.output),
        format: out.format,
        density: out.density,
        status: "ok",
        beforeSize,
        afterSize: fs2.statSync(out.output).size
      });
    } catch (err) {
      try {
        fs2.unlinkSync(out.output + ".tmp");
      } catch {
      }
      results.push({
        source: displaySource,
        output: toDisplayPath(cwd, out.output),
        format: out.format,
        density: out.density,
        status: "error",
        beforeSize,
        afterSize: 0,
        error: err.message
      });
    }
  }
  if (!cfg.dryRun && results.every((r) => r.status === "ok")) {
    cache.record(sourcePath, outPaths, optionsHash);
  }
  if (!cfg.keepOriginal && !cfg.dryRun && results.every((r) => r.status === "ok")) {
    try {
      fs2.unlinkSync(sourcePath);
    } catch {
    }
  }
  return results;
}
async function runParallel(files, cfg, cwd, onProgress) {
  const cache = new ConversionCache(cwd, cfg.cache);
  const concurrency = Math.max(1, Math.min(files.length, os.cpus().length));
  const all = [];
  let index = 0;
  let done = 0;
  const workers = Array.from({ length: concurrency }, async () => {
    while (index < files.length) {
      const i = index++;
      const res = await convertFile(files[i], cfg, cache, cwd);
      all.push(...res);
      done++;
      onProgress?.(done, files.length);
    }
  });
  await Promise.all(workers);
  cache.save();
  return all;
}

// src/scan.ts
import { globby } from "globby";
import path5 from "path";

// src/config.ts
import fs3 from "fs";
import path4 from "path";
var DEFAULT_CONFIG = {
  input: ".",
  output: "",
  recursive: false,
  formats: ["webp"],
  quality: 80,
  resize: null,
  maxWidth: null,
  responsive: [],
  suffix: "",
  keepMetadata: false,
  keepOriginal: true,
  include: [],
  exclude: [],
  cache: true,
  dryRun: false
};
var CONFIG_FILES = ["webimg.config.json", ".webimgrc.json", ".webimgrc"];
function loadConfig(cwd = process.cwd()) {
  for (const name of CONFIG_FILES) {
    const p = path4.join(cwd, name);
    if (fs3.existsSync(p)) {
      try {
        return JSON.parse(fs3.readFileSync(p, "utf8"));
      } catch (e) {
        throw new Error(`Invalid ${name}: ${e.message}`);
      }
    }
  }
  return {};
}
function mergeConfig(...sources) {
  return Object.assign({}, DEFAULT_CONFIG, ...sources);
}

// src/scan.ts
var SUPPORTED_INPUT = ["png", "jpg", "jpeg", "tiff", "webp", "avif"];
async function findImages(cwd, opts = {}) {
  const cfg = mergeConfig(opts);
  const inputRoot = resolveInputRoot(cwd, cfg);
  const ext = `{${SUPPORTED_INPUT.join(",")}}`;
  const base = cfg.recursive ? [`**/*.${ext}`] : [`*.${ext}`];
  const patterns = cfg.include.length ? cfg.include : base;
  const ignore = [
    "**/node_modules/**",
    "**/dist/**",
    "**/.git/**",
    ...cfg.exclude
  ];
  const outputRoot = resolveOutputRoot(cwd, cfg);
  if (outputRoot && isSubpath(inputRoot, outputRoot)) {
    const rel = path5.relative(inputRoot, outputRoot);
    if (rel) ignore.push(`${toPosixPath(rel)}/**`);
  }
  const files = await globby(patterns, { cwd: inputRoot, ignore, caseSensitiveMatch: false, absolute: true });
  const normalized = files.map((f) => toDisplayPath(cwd, path5.normalize(f)));
  return filterPlannedOutputsFromInputs(normalized, cfg, cwd);
}

// src/report.ts
import chalk from "chalk";
import fs4 from "fs";
import path6 from "path";
function formatBytes(bytes) {
  const sign = bytes < 0 ? "-" : "";
  const abs = Math.abs(bytes);
  if (abs < 1024) return `${sign}${abs} B`;
  if (abs < 1024 * 1024) return `${sign}${(abs / 1024).toFixed(1)} KB`;
  return `${sign}${(abs / 1024 / 1024).toFixed(2)} MB`;
}
function formatGain(beforeSize, afterSize, precision = 0) {
  if (!beforeSize) return "\u2014";
  const savedPercent = (beforeSize - afterSize) / beforeSize * 100;
  const sign = savedPercent >= 0 ? "-" : "+";
  return `${sign}${Math.abs(savedPercent).toFixed(precision)}%`;
}
function summarize(results) {
  const ok = results.filter((r) => r.status === "ok");
  const beforeBySource = /* @__PURE__ */ new Map();
  for (const r of ok) {
    if (!beforeBySource.has(r.source)) beforeBySource.set(r.source, r.beforeSize);
  }
  const totalBefore = [...beforeBySource.values()].reduce((s, n) => s + n, 0);
  const totalAfter = ok.reduce((s, r) => s + r.afterSize, 0);
  const saved = totalBefore - totalAfter;
  return {
    total: results.length,
    ok: ok.length,
    skipped: results.filter((r) => r.status === "skipped").length,
    cached: results.filter((r) => r.status === "cached").length,
    errors: results.filter((r) => r.status === "error").length,
    dryrun: results.filter((r) => r.status === "dryrun").length,
    totalBefore,
    totalAfter,
    savedBytes: saved,
    savedPercent: totalBefore > 0 ? saved / totalBefore * 100 : 0,
    results
  };
}
function printReport(s) {
  console.log(`
\u{1F4CA} ${chalk.bold("Rapport")}`);
  console.log("\u2500".repeat(78));
  const ok = s.results.filter((r) => r.status === "ok");
  if (ok.length > 0) {
    console.log(
      chalk.bold(
        "Source".padEnd(34) + "Format".padEnd(8) + "Avant".padEnd(12) + "Apr\xE8s".padEnd(12) + "Gain"
      )
    );
    for (const r of ok) {
      const gain = formatGain(r.beforeSize, r.afterSize);
      const name = r.source.length > 32 ? "\u2026" + r.source.slice(-31) : r.source;
      console.log(
        name.padEnd(34) + `${r.format}${r.density > 1 ? "@" + r.density + "x" : ""}`.padEnd(8) + formatBytes(r.beforeSize).padEnd(12) + formatBytes(r.afterSize).padEnd(12) + chalk.green(gain)
      );
    }
    console.log("\u2500".repeat(78));
    console.log(
      chalk.bold("TOTAL".padEnd(42)) + formatBytes(s.totalBefore).padEnd(12) + formatBytes(s.totalAfter).padEnd(12) + chalk.green.bold(formatGain(s.totalBefore, s.totalAfter, 1))
    );
    console.log(`\u{1F4BE} \xC9conomie : ${chalk.green.bold(formatBytes(s.savedBytes))}`);
  }
  console.log("");
  if (s.ok) console.log(`\u2705 Converties : ${chalk.green(s.ok)}`);
  if (s.cached) console.log(`\u26A1 Depuis le cache : ${chalk.cyan(s.cached)}`);
  if (s.skipped) console.log(`\u23ED\uFE0F  Ignor\xE9es (existantes) : ${chalk.yellow(s.skipped)}`);
  if (s.dryrun) console.log(`\u{1F50D} Dry-run : ${chalk.blue(s.dryrun)} sortie(s) planifi\xE9e(s)`);
  if (s.errors) {
    console.log(`\u274C Erreurs : ${chalk.red(s.errors)}`);
    s.results.filter((r) => r.status === "error").forEach((e) => console.log(`   ${chalk.red("\u2022")} ${e.source} \u2192 ${e.error}`));
  }
}
function writeJsonReport(p, s) {
  fs4.mkdirSync(path6.dirname(p), { recursive: true });
  fs4.writeFileSync(p, JSON.stringify(s, null, 2));
}
function writeMarkdownReport(p, s) {
  const lines = [
    "# webimg-cli report",
    "",
    `- Converties: **${s.ok}**`,
    `- Depuis cache: ${s.cached}`,
    `- Ignor\xE9es: ${s.skipped}`,
    `- Erreurs: ${s.errors}`,
    `- Total avant: ${formatBytes(s.totalBefore)}`,
    `- Total apr\xE8s: ${formatBytes(s.totalAfter)}`,
    `- \xC9conomie: **${formatBytes(s.savedBytes)} (-${s.savedPercent.toFixed(1)}%)**`,
    "",
    "| Source | Format | Avant | Apr\xE8s | Gain |",
    "|---|---|---|---|---|",
    ...s.results.filter((r) => r.status === "ok").map((r) => {
      const gain = formatGain(r.beforeSize, r.afterSize);
      return `| ${r.source} | ${r.format}${r.density > 1 ? "@" + r.density + "x" : ""} | ${formatBytes(r.beforeSize)} | ${formatBytes(r.afterSize)} | ${gain} |`;
    }),
    ""
  ];
  fs4.mkdirSync(path6.dirname(p), { recursive: true });
  fs4.writeFileSync(p, lines.join("\n"));
}
export {
  ConversionCache,
  DEFAULT_CONFIG,
  SUPPORTED_INPUT,
  convertFile,
  findImages,
  formatBytes,
  loadConfig,
  mergeConfig,
  printReport,
  runParallel,
  summarize,
  writeJsonReport,
  writeMarkdownReport
};
